src/details/ArborX_DetailsTreeConstruction.hpp

```diff
@@ -330,221 +200,209 @@ inline void initializeLeafNodes(
-            typename... ParentsViewProperties>
-  CalculateInternalNodesBoundingVolumesFunctor(
-      ExecutionSpace const &space,
-      Kokkos::View<Node *, NodesViewProperties...> internal_and_leaf_nodes,
-      Kokkos::View<int const *, ParentsViewProperties...> parents,
-      size_t n_internal_nodes)
-      : _flags(Kokkos::ViewAllocateWithoutInitializing("flags"),
-               n_internal_nodes)
-      , _parents(parents)
-      , _internal_and_leaf_nodes(internal_and_leaf_nodes)
-  {
-    // Initialize flags to zero
-    Kokkos::deep_copy(space, _flags, 0);
-  }
-
-  KOKKOS_FUNCTION
-  void operator()(int i) const
-  {
-    // Walk toward the root and do process it even though technically its
-    // bounding box has already been computed (bounding box of the scene)
-    while (true)
-    {
-      i = _parents(i);
-
-      // Use an atomic flag per internal node to terminate the first
-      // thread that enters it, while letting the second one through.
-      // This ensures that every node gets processed only once, and not
-      // before both of its children are processed.
-      if (Kokkos::atomic_compare_exchange_strong(&_flags(i), 0, 1))
-        break;
-
-      // Internal node bounding boxes are unitialized hence the
-      // assignment operator below.
-      // FIXME: accessing Node::bounding_box is not ideal but I was
-      // reluctant to pass the bounding volume hierarchy to
-      // generateHierarchy()
-      Node *node = &_internal_and_leaf_nodes(i);
-      Node const *first_child = &_internal_and_leaf_nodes(node->children.first);
-      Node const *second_child =
-          &_internal_and_leaf_nodes(node->children.second);
-      node->bounding_box = first_child->bounding_box;
-      expand(node->bounding_box, second_child->bounding_box);
-      if (i == 0) // root node
-        break;
-    }
-    // NOTE: could check that bounding box of the root node is indeed the
-    // union of the two children.
-  }
-
-private:
-  // Use int instead of bool because CAS (Compare And Swap) on CUDA does not
-  // support boolean
-  Kokkos::View<int *, MemorySpace> _flags;
-  Kokkos::View<int const *, MemorySpace> _parents;
-  Kokkos::View<Node *, MemorySpace> _internal_and_leaf_nodes;
-};
-
-template <typename ExecutionSpace, typename... LeafNodesViewProperties,
-          typename... InternalNodesViewProperties,
-          typename... ParentsViewProperties>
-void calculateInternalNodesBoundingVolumes(
-    ExecutionSpace const &space,
-    Kokkos::View<Node const *, LeafNodesViewProperties...> leaf_nodes,
-    Kokkos::View<Node *, InternalNodesViewProperties...> internal_nodes,
-    Kokkos::View<int const *, ParentsViewProperties...> parents)
-{
-  auto const first = internal_nodes.extent(0);
-  auto const last = first + leaf_nodes.extent(0);
-  Kokkos::View<Node *, InternalNodesViewProperties...,
-               Kokkos::MemoryTraits<Kokkos::Unmanaged>>
-      internal_and_leaf_nodes(internal_nodes.data(), last);
```

Comment:
So, in the previous implementation we did this. It could save us some headache in the new one.